### DISCUSSION TITLE
Upload content not working from UI

### DISCUSSION THREAD

**Discussion by tsprasath** [Sunbird-Ed/Community#318] [Installation] [answered]
Hi team,

CSP:GCP

We are trying to upload content from our workspace, it is not uploading throwing an error.
![Screenshot from 2022-12-23 13-47-58](https://user-images.githubusercontent.com/20708426/209299481-99569ed6-b189-4b20-b922-591d51e47090.png)

**Comment by tsprasath** [comment]
![Screenshot from 2022-12-30 11-44-41](https://user-images.githubusercontent.com/20708426/210238620-70f3d946-6900-4d7f-8c95-59087966dbdd.png)
throws 403 error

**Reply by gandham-santhosh** [reply]
@pallakartheekreddy @maheshkumargangula can you please check this?

**Reply by pallakartheekreddy** [reply]
@tsprasath Could you let me know the following details:
1. What version of Upload content editor using?
2. Which API giving 403 error?

**Reply by tsprasath** [reply]
@pallakartheekreddy  

CSP- Gcloud
content-editor release-5.2.0_RC1, While uploading content like pdf it is unable to upload.

**Reply by Krishnaj20** [reply]
@pallakartheekreddy - can you address this query?

**Reply by pallakartheekreddy** [reply]
@tsprasath Can you share Which API giving 403 error? Above screenshot not showing API details…

**Reply by vaibhavbhuva** [reply]
@pallakartheekreddy I think this issue has been resolved.
@tsprasath Can you please confirm the same ?

**Reply by maheshkumargangula** [reply]
@tsprasath @vaibhavbhuva – Can you please share what is the root cause of the issue and the fix? This will be helpful for other adopters to understand the issue and resolve themselves. 
Cc: @pallakartheekreddy @Krishnaj20

**Comment by tsprasath** [comment, marked as answer]
This issue got resolved

**Comment by sharathkashyap** [comment]
@maheshkumargangula @pallakartheekreddy 
The issue of content upload was an issue in Knowledge platform not able to create a presigned url correctly using gcp key and secret. 
Root Cause: GCP has introduced a new v4 signing for generating signed url. However, the library used in the cloud storage sdk
`  <dependency>
			<groupId>org.apache.jclouds</groupId>
			<artifactId>jclouds-all</artifactId>
			<version>2.2.1</version>
        </dependency>`
This repo do not have the updated mechanism of generating presigned url. There was a logical extension written and contributed back to cloud sdk in below PR:
`https://github.com/project-sunbird/sunbird-cloud-storage-sdk/pull/45`

There are also changes to be done on content-plugins. The PR will be raised shortly.

**Reply by sharathkashyap** [reply]
@maheshkumargangula @pallakartheekreddy We see that PR has not been reviewed yet by the larger forum. Could you please review the PR for the contribution.

cc: @tsprasath

**Reply by maheshkumargangula** [reply]
@tsprasath @sharathkashyap – Can you please tag the JIRA ticket or the discussion thread and explain what is the issue we have fixed with the PR?

**Reply by sharathkashyap** [reply]
@maheshkumargangula This entire PR has been raised in a reference to present discussion forum thread. I have clearly explained the issue it is trying to solve above.

cc: @tsprasath

**Reply by mathewjpallan** [reply]
The current approach in the cloud-storage SDK was only generating http endpoints and this would not work for XHR requests from a web page loaded via https. Using the v4 API provided by GCP does return an https signed URL and this is what the PR is about. This change has been tested on GCP. @maheshkumargangula can you please review the PR
